### PR TITLE
[RHELC-664] Collect breadcrumbs through RHSM custom facts

### PR DESCRIPTION
**Pull request by r0x0d**
This PR introduces the collection of breadcrumbs through subscription-manager, by actually converting the current state of our breadcrumbs into RHSM custom facts and uploading it to the candlepin server tied to the registration. 

Currently, we are introducing this as an MVP to the general idea. In the future, we will stop outputting the file at `/etc/migration-results` and stick only with the one that is located at `/etc/rhsm/facts/*.facts`, but for now, we have both of them placed on the system in case of something going wrong and not having the facts synced with subscription-manager.

### TODO
- [ ] Create tasks on jira to keep improving the existing breadcrumbs collection
- [ ] TODO
 

Jira Issue: [RHELC-664](https://issues.redhat.com/browse/RHELC-664)

Checklist
- [x] PR meets acceptance criteria specified in the Jira issue
- [x] PR has been tested manually in a VM (either author or reviewer)
- [ ] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending`

### CODE REVIEW

File: convert2rhel/systeminfo.py
Comment (r0x0d) [inline convert2rhel/systeminfo.py:439]: A note about how we get the version here… Using the `self._get_system_version()` function, we will lose the patch number in a version, for example: 

```
7.6.1810 -> 7.6
```
The `.1810` part will be cut off, as it returns an instance of the class `Version` that currently only matches the first two part. If that's not a problem, I would say we can continue with this and change in the future, or, we could go ahead in this PR and adapt the `Version` class to accept one more version (`Version.patch`) and mark it as optional.

**Review by bocekm — DISMISSED** [review]

File: convert2rhel/breadcrumbs.py
Comment (abadger) [inline convert2rhel/breadcrumbs.py:196]:
```suggestion
        # Try to read the contents of the file, if that fails, assign the
```

File: convert2rhel/breadcrumbs.py
Comment (abadger) [inline convert2rhel/breadcrumbs.py:194]: We aren't currently saving any secrets in the files that I can see but since we take the time to make the permissions restrictive, the chmod should be done here.  That way the permissions are set to 0o600 before we write anything new to the file.

File: convert2rhel/breadcrumbs.py
Comment (r0x0d) [reply convert2rhel/breadcrumbs.py:196]: done.

File: convert2rhel/subscription.py
Comment (r0x0d) [reply convert2rhel/subscription.py:995]: done.

File: convert2rhel/utils.py
Comment (r0x0d) [reply convert2rhel/utils.py:584]: done.

File: convert2rhel/breadcrumbs.py
Comment (r0x0d) [reply convert2rhel/breadcrumbs.py:212]: Kinda of… I'm reusing this piece of code to have a single write at the end of it, but changing in a case where we don't have a `key`. 

Now that I'm thinking more about it, maybe should be a good idea to split those functions as something like: 

```python
# utils.py
def write_json_to_file(path, data, mode): ...

def write_json_to_file_with_key(path, data, key): 
    # somewhere in the workflow 
    write_json_to_file(path, data)
 ```
 
 Not exactly sure... Do you have any suggestions to re-write this method, @abadger?

Comment (r0x0d) [reply convert2rhel/breadcrumbs.py:215]: Maybe we can tackle this in our 1:1 today

Comment (abadger) [reply convert2rhel/breadcrumbs.py:215]: Your idea of creating two functions sounds good to me.  If you're going to move them into urils, I think I'd have `mode` be a parameter to both functions.

File: convert2rhel/breadcrumbs.py
Comment (r0x0d) [reply convert2rhel/breadcrumbs.py:199]: removed that comment.

File: convert2rhel/breadcrumbs.py
Comment (abadger) [inline convert2rhel/breadcrumbs.py:192]: Let's do the chmod everytime, that way if the file already exists but with relaxed permissions, we'll be sure to make it secure.
```suggestion
    # the file can be changed just by root
    os.chmod(path, 0o600)
```

File: convert2rhel/utils.py
Comment (abadger) [inline convert2rhel/utils.py:649]: Oh, I thought mode would be the permissions.  So, `0o600`, for instance.

File: convert2rhel/utils.py
Comment (abadger) [inline convert2rhel/utils.py:649]:
```suggestion
def write_json_object_to_file(path, data, mode="0o600"):
```

File: convert2rhel/utils.py
Comment (abadger) [inline convert2rhel/utils.py:659]:
```suggestion
    with open(path, mode="w") as handler:
```

File: convert2rhel/utils.py
Comment (abadger) [inline convert2rhel/utils.py:661]:
```suggestion
        os.chmod(path, mode)
        json.dump(data, handler, indent=4)
```

File: convert2rhel/utils.py
Comment (abadger) [inline convert2rhel/utils.py:656]:
```suggestion
    :param mode: The permissions for the file.
```

**Comment by r0x0d** [comment]
@abadger since I'm using a couple of internal functions of SystemInfo class, should we make it public?

File: convert2rhel/utils.py
Comment (r0x0d) [reply convert2rhel/utils.py]: done.

**Comment by abadger** [comment]
The code looks good.  Marking this dev complete.

**Review by abadger — DISMISSED** [review]

**Comment by danmyway** [comment]
Verification looks good to me! Thank you, @r0x0d!

**Review by danmyway — DISMISSED** [review]

**Comment by abadger** [comment]
Rodolfo, Ill let you merge this so you can decide how much you want to squash.

File: convert2rhel/systeminfo.py
Comment (bocekm) [inline convert2rhel/systeminfo.py:160]: Personally I don't think anyone would need this "id" when collecting the breadcrumbs for statistics. "Final" or "Core" doesn't even resemble any kind of identification. In RHEL the parentheses contain a codename, e.g. Ootpa. CentOS Stream has no parenthesis the same as Oracle Linux ("CentOS Stream release 8").
For that reason, I'd remove this function together with publishing this id in the breadcrumbs.

Comment (r0x0d) [reply convert2rhel/systeminfo.py:160]: This is mostly relevant for `target_os` rather than the `source_os`, indeed, as the codename in RHEL is more important on statistics, I would say we could leave this as it is right now as it won't affect anything or anyone if `source_os` has an extra `id` there (also, making it more “complete”). 

Maybe I misunderstood @tabowling example when we were talking about this feature, probably he wanted the `id` only for the `target_os`, but honestly, I don't see a problem with both of them having the id... Most of the time I think this would be empty for `source_os`.

**Review by abadger — APPROVED** [review]
Assuming that the integration test failures that are relevant for this change are now fixed, this is approved by both Dan (QE) and myself (DEV)

**Comment by danmyway** [comment]
> Assuming that the integration test failures that are relevant for this change are now fixed, this is approved by both Dan (QE) and myself (DEV)

Some of those are still failing with the following, I will try to see, what the problem might be locally, @abadger 
```
=================================== FAILURES ===================================
________________________ test_flag_system_as_converted _________________________

    def test_flag_system_as_converted():
        """Test if the breadcrumbs file was created and corresponds to the JSON schema."""
    
        assert os.path.exists(C2R_MIGRATION_RESULTS)
>       assert os.path.exists(C2R_RHSM_CUSTOM_FACTS)
E       AssertionError: assert False
E        +  where False = <function exists at 0x7f7d8da4cc80>('/etc/rhsm/facts/convert2rhel.facts')
E        +    where <function exists at 0x7f7d8da4cc80> = <module 'posixpath' from '/usr/lib64/python3.6/posixpath.py'>.exists
E        +      where <module 'posixpath' from '/usr/lib64/python3.6/posixpath.py'> = os.path
```

**Comment by abadger** [comment]
Note:  source_os is set to "null" sometimes when we fail.  We had not thought that should happen when we discussed the issue on Friday but when I got deep into why that was happening on some of the integration tests here, I found that those integration tests are failing before `early_collection()` is run to get the information about `source_os`.

I think we should talk about whether to run "early_collection" in rollback if it hasn't been done already but I think that's should be done in a separate issue.

**Comment by abadger** [comment]
The latest run had three errors:
### One
> E               0: re.compile("WARNING - The system has \d+ packages not updated.")

Every run had this one.  It looks like it was already present in main and should be fixed by this recent change: 999c178ff0c3483e6ee9d9058c55cca3ab8a99d8   I've rebased to try fixing it.

### Two
> E           On instance['activities'][0]['source_os']:
> E               'null'
>
> /usr/lib/python3.6/site-packages/jsonschema/validators.py:130: ValidationError

This was only present on centos8.  It is what https://github.com/oamg/convert2rhel/pull/612/commits/0c6ae68140ef87d534446ef7151860e549065f7b was intended to fix and that change appears to have fixed the problem on every other platform so I'm not sure why the test is still failing here.  It's as if the centos8 build did not pick up that change to the schema file used in the test?

### Three
> 01:10:19         out:  Problem 1: cannot install the best update candidate for package gcc-8.5.0-10.1.0.1.el8_6.x86_64
> 01:10:19         out:   - nothing provides libgomp = 8.5.0-10.1.el8_6 needed by gcc-8.5.0-10.1.el8_6.x86_64
> 01:10:19         out:  Problem 2: cannot install the best update candidate for package platform-python-devel-3.6.8-47.0.1.el8_6.x86_64
> 01:10:19         out:   - nothing provides python3-libs(x86-64) = 3.6.8-47.el8_6 needed by platform-python-devel-3.6.8-47.el8_6.x86_64
> 01:10:19         out:  Problem 3: package annobin-10.29-3.el8.x86_64 requires (gcc >= 8 with gcc < 9), but none of the providers can be installed
> 01:10:19         out:   - package gcc-8.3.1-5.1.el8.x86_64 requires cpp = 8.3.1-5.1.el8, but none of the providers can be installed
> 01:10:19         out:   - package gcc-8.3.1-4.5.el8.x86_64 requires cpp = 8.3.1-4.5.el8, but none of the providers can be installed
> 01:10:19         out:   - package gcc-8.4.1-1.el8.x86_64 requires cpp = 8.4.1-1.el8, but none of the providers can be installed
> 01:10:19         out:   - package gcc-8.3.1-5.el8.x86_64 requires cpp = 8.3.1-5.el8, but none of the providers can be installed
> 01:10:19         out:   - package gcc-8.2.1-3.5.el8.x86_64 requires cpp = 8.2.1-3.5.el8, but none of the providers can be installed
> 01:10:19         out:   - package gcc-8.5.0-3.el8.x86_64 requires cpp = 8.5.0-3.el8, but none of the providers can be installed
> 01:10:19         out:   - package gcc-8.5.0-4.el8_5.x86_64 requires cpp = 8.5.0-4.el8_5, but none of the providers can be installed
> 01:10:19         out:   - package gcc-8.5.0-10.el8.x86_64 requires cpp = 8.5.0-10.el8, but none of the providers can be installed
> 01:10:19         out:   - cannot install both cpp-8.5.0-10.1.el8_6.x86_64 and cpp-8.3.1-5.1.el8.x86_64
> 01:10:19         out:   - cannot install both cpp-8.5.0-10.1.el8_6.x86_64 and cpp-8.3.1-4.5.el8.x86_64
> 01:10:19         out:   - cannot install both cpp-8.5.0-10.1.el8_6.x86_64 and cpp-8.4.1-1.el8.x86_64
> 01:10:19         out:   - cannot install both cpp-8.5.0-10.1.el8_6.x86_64 and cpp-8.3.1-5.el8.x86_64
> 01:10:19         out:   - cannot install both cpp-8.5.0-10.1.el8_6.x86_64 and cpp-8.2.1-3.5.el8.x86_64
> 01:10:19         out:   - cannot install both cpp-8.5.0-10.1.el8_6.x86_64 and cpp-8.5.0-3.el8.x86_64
> 01:10:19         out:   - cannot install both cpp-8.5.0-10.1.el8_6.x86_64 and cpp-8.5.0-4.el8_5.x86_64
> 01:10:19         out:   - cannot install both cpp-8.5.0-10.1.el8_6.x86_64 and cpp-8.5.0-10.el8.x86_64
> 01:10:19         out:   - cannot install the best update candidate for package cpp-8.5.0-10.1.0.1.el8_6.x86_64
> 01:10:19         out:   - cannot install the best update candidate for package annobin-10.29-3.0.1.el8.x86_64
> 01:10:19         out:   - gcc-8.5.0-10.1.0.1.el8_6.x86_64 does not belong to a distupgrade repository
> 01:10:19         out:   - nothing provides libgomp = 8.5.0-10.1.el8_6 needed by gcc-8.5.0-10.1.el8_6.x86_64

This yum error is only happening on oracle-8.6.  It is an error resolving packages to yum which is code that this PR does not touch.  This might be a problem specific to the packages available in the  oracle 8 repositories and have nothing to do with this PR.

**Comment by abadger** [comment]
Figured out Two: source_os is being set to null but the fix was allowing source_os:id to be set to null.  I've updated the schema so that should pass too.

**Comment by abadger** [comment]
Okay, most integration tests run via pakit have now completed successfully.  The yum issue is still there but on this retry, it was on centos-8.4 and oraclelinux-8.6. (oraclelinux-7 has not completed although it was started 5 hours ago... I think the yum_distro_sync test there has gotten stuck in RUNNING state.  Every other test has passed.).

Since the problem is occurring on both centos-8.4 and oraclelinux-8.6, I don't think it's a problem with the distribution but I don't think it's being caused by this code either.... Maybe it's a problem that snuck into the codebase when we were having so many false positives in integration testing for a while.

I think this is dev complete, but I'll wait for @danmyway decide if we should merge or if he thinks the yum problems are caused by this.

**Comment by abadger** [comment]
Hmm.. Update:  I looked closer and there are two different errors:

### Centos 84:
network error when  subscription-manager tries to attach a subscription to the machine:
> 12:21:39         out: [10/21/2022 12:21:39] DEBUG - Calling command 'subscription-manager attach --pool 8a99f9af7c95644d017cbb877eec04f8'
> 12:22:26         out: Network error, unable to connect to server. Please see /var/log/rhsm/rhsm.log for more information.
> 12:22:26         out: CRITICAL - Unsuccessful attachment of a subscription.

Leads to test_yum_distrosync failing.

Seems unrelated to breadcrumbs.

### Oraclelinux 8.6
This is the yum error I mentioned above.

Since it is only happening on oraclelinux 8.6, it could be a problem at the distro packaging level... Have to test to find out.

Causes test_run_conversion_using_custom_repos to fail

Also seems unrelated to breadcrumbs changes.

**Review by abadger — APPROVED** [review]
Approved by Dan and myself.

**Comment by bocekm** [comment]
@r0x0d, one suggestion for a subsequent improvement - the output related to breadcrumbs:

- is printed under a previous task (`Final: Remove temporary folder`),
- there's a newline missing (`Release Info: id: Ootpa name: Red Hat Enterprise Linux version: 8.6`),
- the "migration logs" are unreadable.

You can either remove the migration logs as they have little value to the user, or leave there just one (the "updated" one) but pretty-print it.

Screenshot from 2022-10-28 00-31-03

**Comment by bocekm** [comment]
Also, the name and version is printed three times when gathering system info:
```
[10/27/2022 22:35:54] TASK - [Prepare: Gather system information] *******************************
Name:                CentOS
OS version:          6.10
[10/27/2022 22:35:54] DEBUG - Calling command 'uname -i'
Architecture:        x86_64
Config filename:     centos-6-x86_64.cfg
Skipping the execution of 'rpm -Va'.
[10/27/2022 22:35:54] DEBUG - Calling command 'uname -r'
[10/27/2022 22:35:54] DEBUG - Booted kernel VRA (version, release, architecture): 2.6.32-754.35.1.el6.x86_64
Checking internet connectivity using address 'http://static.redhat.com/test/rhel-networkmanager.txt'.
Successfully connected to address 'http://static.redhat.com/test/rhel-networkmanager.txt', internet connection seems to be available.
[10/27/2022 22:35:54] DEBUG - Calling command '/sbin/service messagebus status'
messagebus is stopped
[10/27/2022 22:35:55] DEBUG - Calling command '/sbin/service messagebus status'
messagebus is stopped
[10/27/2022 22:35:57] DEBUG - Calling command '/sbin/service messagebus status'
messagebus is stopped
Name:                CentOS
OS version:          6.10
Release Info: version: 6.10 id: Final name: CentOS
```

File: convert2rhel/breadcrumbs.py
Comment (bocekm) [inline convert2rhel/breadcrumbs.py:184]: I think that if the folder /etc/rhsm/facts/ does not exist, we should create it before trying to write the facts file to it.

The way it is now, we will have the breadcrumbs file available only after we install the subscription-manager package. But we'd like to know about even about the failures before we install the subscription-manager package. As a matter of fact, the subscription-manager pkg is not even being installed when the user runs the tool with the `--norhsm` option.

The facts are supposed to be collected by sos. We should check whether the sos plugin collects them even without the subscription-manager installed.

<details>
  <summary>Example of where the breadcrumbs should have been written</summary>

```
[10/27/2022 22:39:25] TASK - [Prepare: Checking if the loaded kernel version is the most recent] 
[10/27/2022 22:39:25] DEBUG - Calling command 'repoquery --quiet --qf "%{BUILDTIME}\t%{VERSION}-%{RELEASE}\t%{REPOID}" kernel'
[10/27/2022 22:39:25] DEBUG - Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/convert2rhel/main.py", line 94, in main
    checks.perform_pre_checks()
  File "/usr/lib/python2.6/site-packages/convert2rhel/checks.py", line 68, in perform_pre_checks
    is_loaded_kernel_latest()
  File "/usr/lib/python2.6/site-packages/convert2rhel/checks.py", line 672, in is_loaded_kernel_latest
    _, latest_kernel, repoid = packages[0]
ValueError: need more than 1 value to unpack

Writing breadcrumbs to '/etc/migration-results'.
[10/27/2022 22:39:25] DEBUG - Prior migration log: {"activities": [{"executed": "/usr/bin/convert2rhel -y --debug --no-rpm-va --serverurl subscription.rhsm.stage.redhat.com --username c2r_main --password ***** --pool 2c94a05683f64f030184162087bf18cb", "success": false, "run_id": "null", "activity_ended": "2022-10-27T20:36:11.097896Z", "activity_started": "2022-10-27T20:36:01.609912Z", "version": "1", "env": {}, "activity": "conversion", "source_os": {"version": "6.10", "id": "Final", "name": "CentOS"}, "packages": [{"nevra": "convert2rhel-1.0-1.20221027201547923673.pr587.61.gda871a6.el6.noarch", "signature": "RSA/8, Thu Oct 27 22:17:20 2022, Key ID 175e6797ecf2dde9"}], "target_os": "null"}]}
[10/27/2022 22:39:25] DEBUG - Updated migration log: {"activities": [{"executed": "/usr/bin/convert2rhel -y --debug --no-rpm-va --serverurl subscription.rhsm.stage.redhat.com --username c2r_main --password ***** --pool 2c94a05683f64f030184162087bf18cb", "success": false, "run_id": "null", "activity_ended": "2022-10-27T20:36:11.097896Z", "activity_started": "2022-10-27T20:36:01.609912Z", "version": "1", "env": {}, "activity": "conversion", "source_os": {"version": "6.10", "id": "Final", "name": "CentOS"}, "packages": [{"nevra": "convert2rhel-1.0-1.20221027201547923673.pr587.61.gda871a6.el6.noarch", "signature": "RSA/8, Thu Oct 27 22:17:20 2022, Key ID 175e6797ecf2dde9"}], "target_os": "null"}, {"executed": "/usr/bin/convert2rhel -y --debug --no-rpm-va --serverurl subscription.rhsm.stage.redhat.com --username c2r_main --password ***** --pool 2c94a05683f64f030184162087bf18cb", "run_id": "null", "packages": [{"nevra": "convert2rhel-1.0-1.20221027201547923673.pr587.61.gda871a6.el6.noarch", "signature": "RSA/8, Thu Oct 27 22:17:20 2022, Key ID 175e6797ecf2dde9"}], "target_os": "null", "success": false, "activity_ended": "2022-10-27T20:39:25.381003Z", "version": "1", "env": {}, "activity": "conversion", "source_os": {"version": "6.10", "id": "Final", "name": "CentOS"}, "activity_started": "2022-10-27T20:39:19.906445Z"}]}
Writing RHSM custom facts to '/etc/rhsm/facts/convert2rhel.facts'.
WARNING - Unable to find RHSM facts folder at '/etc/rhsm/facts'.
No changes were made to the system.
```
</details>

**Pull request merged**
